### DISCUSSION TITLE
Use multiple tables to define a graph

### DISCUSSION THREAD

**Discussion by dberardo-com** [cwida/duckpgq-extension#161] [Q&A] [answered]
I am looking for an alternative to Apache AGE and I am evaluating this extension.

I will be soon posting a query here to ask if there is an equivalent already available in duckpg but for the time being I would like to understand:

1. Is it possible to define a graph using multiple tables for both vertex and edges, and if so can the same table be used multiple times with different label keys? (I ask this because such example is missing in the doc)
2. Is it necessary to update/refresh the graph table whenever any of the underlying SQL tables have changed? Or is this done automatically by the duckpg engine? If not, is there an incremental way to trigger this update or is it necessary to always drop qnd recreate the graph table?

Thanks

**Comment by Dtenwolde** [comment]
Hi there, sorry for not seeing this discussion thread earlier!

1.	Yes, that’s possible as long as the labels are unique. For instance, you can use the same table with different labels like this:
```sql
CREATE TABLE person (id bigint);
CREATE PROPERTY GRAPH g
VERTEX TABLES (person LABEL person, person LABEL student);
```
This approach lets you define different roles or types for the same table in the graph.

2.	Property graphs in DuckPGQ only need updating if the underlying table or column names change (similar to how a VIEW behaves). Currently, there’s no incremental update mechanism for property graphs. To refresh a graph, you would need to use DROP and CREATE again, or CREATE OR REPLACE PROPERTY GRAPH if you prefer a one-step update.

Hope this helps! Let me know if you have further questions.

**Reply by dberardo-com** [reply]
thanks for the reply.

1. ok, i was expecting a somewhat different looking code snippet to define multiple nodes and edge labels but i will give it a try later. basically what i am trying to achieve is to build a relationship graph between users and resource via user groups and resource groups associations, which in apache AGE would respond to this query sequence: https://github.com/apache/age/issues/2092#issuecomment-2340044202 do you believe this is currently achievable in duckpgq ?
2. you say "similar to how a VIEW behaves"  and then "To refresh a graph, you would need to use DROP and CREATE again" this is somewhat contraddictory, i mean, VIEWs do not contain their own data, so they are always "up-to-date" with the underlying data table(s) because well, they are always re-executed. If that's the case, then we are good, adding more rows to the base tables will automatically "add" those rows' labels to the node/edges graphs. i don't get the second part of your answer here. are you referring only to schema changes (such as adding / renaming columns, etc?)

**Comment by Dtenwolde** [comment, marked as answer]
Hi there, 

Defining a property graph corresponding to the one in the issue you linked is possible. 
[access_control_graph.zip](https://github.com/user-attachments/files/17700032/access_control_graph.zip)
This ZIP contains the data from that issue and schema. In DuckDB (preferably the latest stable version 1.1.3) you can import the database using `import database 'path/to/unzipped/folder'`. 
It contains the following schema: 
```sql
CREATE TABLE belongs_to(operatorid VARCHAR, groupsid VARCHAR);
CREATE TABLE "groups"("name" VARCHAR, _type VARCHAR, "role" VARCHAR);
CREATE TABLE has_permission(groupid VARCHAR, resourceid VARCHAR);
CREATE TABLE "operator"("name" VARCHAR);
CREATE TABLE parent_of(groupid1 VARCHAR, groupid2 VARCHAR);
CREATE TABLE resource("name" VARCHAR, _type VARCHAR);
CREATE TABLE responsible_for(groupid1 VARCHAR, groupid2 VARCHAR);
CREATE TABLE __duckpgq_internal(property_graph VARCHAR, table_name VARCHAR, "label" VARCHAR, is_vertex_table BOOLEAN, source_table VARCHAR, source_pk VARCHAR[], source_fk VARCHAR[], destination_table VARCHAR, destination_pk VARCHAR[], destination_fk VARCHAR[], discriminator VARCHAR, sub_labels VARCHAR[]);
```

And the property graph is defined as: 
```sql
create or replace property graph access_control_graph
    VERTEX TABLES (operator, groups label g, resource)
    EDGE TABLES (parent_of 			SOURCE KEY (groupid1) REFERENCES groups (name)
                         			DESTINATION KEY (groupid2) REFERENCES groups (name),
                 responsible_for 	SOURCE KEY (groupid1) REFERENCES groups (name)
                                 	DESTINATION KEY (groupid2) REFERENCES groups (name),
                 has_permission 	SOURCE KEY (groupid) REFERENCES groups (name)
                    				DESTINATION KEY (resourceid) REFERENCES resource (name),
                 belongs_to	 		SOURCE KEY (operatorid) REFERENCES operator (name)
                    				DESTINATION KEY (groupsid) REFERENCES groups (name));
```

Then with `install duckpgq from community` and `load duckpgq` you should be able to use write PGQ queries, such as: 
```sql
FROM GRAPH_TABLE (access_control_graph MATCH (n:operator)-[r:belongs_to]->(m:g));
```

The second query in your issue mentions `OPTIONAL` matches, which is not yet supported in DuckPGQ see #112. 

2. Sorry I wasn't quite clear. You are correct with: "adding more rows to the base tables will automatically "add" those rows' labels to the node/edges graphs."
For any schema changes in the underlying tables you have to recreate the property graph to reflect those changes. 

I hope that helps :)

**Reply by dberardo-com** [reply]
thanks a lot for your reply and also for showing the "import database" trick of duckdb that i was not aware of.

here my comments on the points of discussion:

1. i can confirm that the graph corresponds to the one in the example on the AGE project (apart from the mistake that Operators should have a role, but not groups). I want to implement some RBAC and ABAC access permission logics using graph based engines thus the need of using OPTIONAL matches, as those might be useful to add "paths" to the query which might be available for Admins, but not for regular Users (Operators in my case). The ABAC part is instead quite straight forward given the possibility to use WHERE statements on both nodes and edges. I hope this use case of mine is understandable.

In this regard i would like to pose 4 questions: 

* why do you have to define "label g" fro the groups table in the graph ?
* is the usage of the Cypher pipe "|" operator as in "MATCH (wallstreet {title: 'Wall Street'})<-[:ACTED_IN|DIRECTED]-(person)" https://github.com/apache/age/issues/2091 available in duckpg ?
* is it possible to replace to "OPTIONAL" matches operators with a concatenation of queies using the standard sql "UNION" operator or is it perhaps not the same thing (i suppose it is not)
* also: is the  Kleene star * operator [ref page 9 of https://homepages.cwi.nl/~boncz/msc/2022-DanielTenWolde.pdf] in (cypher "-[:RESPONSIBLE_FOR*]->") already implemented ?

2. i think we misunderstood each other here. indeed i tried your example and then added (INSERT) a row on both the group and operator tables an rerun the MATCH query and found out that the modifications are applied. This is great!! from the description of #112  i understand that all graphs in duckpg are in fact standard SQL views, meaning that the graphs dont need to be refreshed when new rows are added. This is what i am looking for, but perhaps you might consider adding the possibility to save graphs as MATERIALIZED VIEWS instead of just VIEWS for sake of performance. i guess this might be a optional parameter in the CREATE * queries of the graph

I am going to try this whole material on a more production-like dataset of mine and see how that performs but all this sounds quite promising already

**Reply by dberardo-com** [reply]
i have also found a bug, reported here: https://github.com/cwida/duckpgq-extension/issues/165

**Reply by Dtenwolde** [reply]
Thanks for the bug report! I'll look into this next week and hope to file a bug fix :) 

Supporting the OPTIONAL operator is something I want to implement short term, but I can't make any promises for an exact timeframe. I am also focusing on research and only have limited time, I hope you understand :)

Regarding your questions: 
1) I ran into an issue with just using the table name groups when using the PGQ syntax (see screenshot). I think it's because `groups` is some reserved or unreserved keyword, but I need to investigate further to confirm. The `label g` was a workaround. 
![image](https://github.com/user-attachments/assets/719fca56-32b8-4365-b707-e8b31667ffe2)
2) Does the `|` operator act as an `OR` statement? If so, that's something I want to implement in the long term, but I will not give guarantees at the moment. I think a workaround would be to create a new table which is a `UNION_ALL` on the  `ACTED_IN` and `DIRECTED` tables and use that. 
3) I think the `OPTIONAL` match would translate more to a `LEFT OUTER JOIN` in SQL where all rows on the left side of the join appear at least once (see https://www.w3schools.com/sql/sql_join_left.asp) 
4) The Kleene star is implemented, just keep in mind to use a space between `->` and `*` (see https://github.com/cwida/duckpgq-extension/issues/146). It's a parser error, but this will be fixed soon (see https://github.com/cwida/duckdb-pgq/pull/205). There's more information on the path-finding on the [documentation page](https://www.notion.so/duckpgq/SQL-PGQ-5016889634db44c99cd607205bfbf68a?pvs=4#a61984ba124c44f59e8f40a7cf8f3e57). 

Regarding the materialized property graph, that could be an interesting addition, though that's currently not trivial. I'll keep it in mind :)

Let me know how your implementation goes and if you have any further questions! 
Kind regards, 
Daniel

**Reply by dberardo-com** [reply]
> Supporting the OPTIONAL operator is something I want to implement short term, ...

will keep an eye on that, for that i might start testing without it :)

> 1. I ran into an issue with just using the table name groups

sure i thought the same, just wanted to make sure that was the reason 

> 2. Does the | operator act as an OR statement?

yes, i thought about this workaround too, i think the problem is to deal with more than one such statement at the same time, as this will quickly make the number of UNION ALL queries explode (permutations, i guess). I guess this could be tracked here: https://github.com/cwida/duckpgq-extension/discussions/167

> 3.  think the OPTIONAL match would translate more to a LEFT OUTER JOIN in SQL 

gotcha

> 4. The Kleene star is implemented, just keep in mind to use a space between -> and *

i see, i was not able to relate that to the "path finding" doc page :D

> Regarding the materialized property graph, that could be an interesting addition, though that's currently not trivial. I'll keep it in mind :)

Sure, i watched the youtube video where you mention the CBS, and the last question from the audience goes in this direction. Again, this is not needed for my use case "authorization", but might be useful for some. Btw, in general authorization using graph based queries is kind of an hot topic since the release of google's Zanzibar paper in 2019: https://www.osohq.com/learn/google-zanzibar but current open source solutions are not so well integrable to existing systems in my opinion, and are also too heavy weight when compared with apache AGE or  duckdb. Also they are very specialized in doing just authorization, and this is for me a huge drawback, as it would be nice to have a fully fledged graphdb at hand rather than one that can only deal with auth queries. Here the most relevant example i could find of such tech: https://github.com/authzed/spicedb  
Maybe you were already aware of this branch, but i just wanted to stress the point the graph query engines are not just "hot" in the context of social media analytics, but also beyond that.

Cheers

**Comment by dberardo-com** [comment]
also: to "close the implementation circle" in my use case, i still need to synchronize the underlying tables living in a postgresql database with duckdb so that they can be worked out in duckpgq. i have open a discussion here as this goes beyond the topic at hand: https://github.com/cwida/duckpgq-extension/discussions/166
